### ISSUE TITLE
Implicit relationship between the names of a module and the source file that contains it assumed

### ISSUE THREAD

**Issue by arjenmarkus** [fortran-lang/fpm#128]
I have run into a problem that is illustrated with the attached project:

`
   $ d:/fortran/fpm/fpm.exe build
   # gfortran (for build/gfortran_debug/mxmy/mx.o build/gfortran_debug/mxmy/mx.mod)
   fpm.exe: Error when running Shake build system:
      at want, called at src\Build.hs:205:11 in fpm-0.1.0.0-9a29EtwrI5D7C1C0BiAJeV:Build
   * Depends on: build/gfortran_debug/mxmy/libmxmy.a
      at need, called at src\Build.hs:203:13 in fpm-0.1.0.0-9a29EtwrI5D7C1C0BiAJeV:Build
   * Depends on: build/gfortran_debug/mxmy/mx.o
      at &%>, called at src\Build.hs:183:11 in fpm-0.1.0.0-9a29EtwrI5D7C1C0BiAJeV:Build
   * Depends on: build/gfortran_debug/mxmy/mx.o build/gfortran_debug/mxmy/mx.mod
      at error, called at src\Development\Shake\Internal\Rules\Files.hs:245:13 in shake-   0.18.5-7VS8ovPeUeSCopHgkQQ1Vu:Development.Shake.Internal.Rules.Files
   * Raised the exception:
   Error, &%> rule failed to produce 1 file (out of 2)
     build/gfortran_debug/mxmy/mx.o
     build/gfortran_debug/mxmy/mx.mod - MISSING
`

The source file is called "mx.f90" but it contains a module "my". Apparently this is not currently supported, whether by design or otherwise. It seems to me to be a bug. 
[test-fpm-mxmy.zip](https://github.com/fortran-lang/fpm/files/4952475/test-fpm-mxmy.zip)

**Comment by milancurcic** [comment]
Hi Arjen, yes, it's by design, see https://github.com/fortran-lang/fpm/blob/master/PACKAGING.md#single-module-library.

The rationale is that if we have higher constraints on the user application or library structure, we can design a more robust build system that has a "sane default" with fewer edge cases.

Do you think the packaging guide could be improved in any way?

**Comment by milancurcic** [comment]
However, the unhelpful response from fpm is not by design, but is due to this exception handling not being implemented. In scenarios like this fpm must be helpful to the user, for example:

```
fpm error: I noticed that you defined a module `my` in file `src/mx.f90`. The module name must match the source (path) name. Either rename your module to `mx`, or place it in `src/my.f90`.
```

or similar.

**Comment by certik** [comment]
Yes, the default layout that fpm expects is on purpose restricted, similar to how Cargo does it. The module names must be consistent with the filename.

However, I am all for allowing exceptions, which you could specify by hand in `fpm.toml`. We should do that.

The error message is not helpful. But we should work on fixing both of these things in the Fortran implementation (see #124).

**Comment by arjenmarkus** [comment]
It is not criticism :) - just a user experience. I understand the
limitations, but an innocent user may run into this sort of things and not
be able to pinpoint the cause (as I was not with my other problem).

Regards,

Arjen

Op di 21 jul. 2020 om 16:56 schreef Ondřej Čertík <notifications@github.com

**Comment by milancurcic** [comment]
@certik I opened #129 and #130 to track these specific exception handling and error message issues.

I also created the fpm-fortran tag so that we can tag any issue to be specific to the Fortran implementation and not Haskell. You'll notice that I didn't initially tag #129 and #130 as fpm-fortran, as I'm unsure whether we should leave these unattended for a long time before fpm-fortran is up to speed. I think they're quite critical UI issues.

**Comment by certik** [comment]
And I implemented the initial infrastructure for Fortran at #131 to start getting things moving.

**Comment by awvwgk** [comment]
This might be fixed with #213.

**Comment by everythingfunctional** [comment]
The bootstrap version is no longer enforcing any match between module and source file name.
